core/napi.rs

```diff
@@ -0,0 +1,502 @@
+  napi_enumerable | napi_configurable | napi_writable;
+
+#[repr(C)]
+#[derive(Copy, Clone, Debug)]
+pub struct napi_property_descriptor {
+  pub utf8name: *const c_char,
+  pub name: napi_value,
+  pub method: napi_callback,
+  pub getter: napi_callback,
+  pub setter: napi_callback,
+  pub value: napi_value,
+  pub attributes: napi_property_attributes,
+  pub data: *mut c_void,
+}
+
+#[repr(C)]
+#[derive(Debug)]
+pub struct napi_extended_error_info {
+  pub error_message: *const c_char,
+  pub engine_reserved: *mut c_void,
+  pub engine_error_code: i32,
+  pub status_code: napi_status,
+}
+
+#[repr(C)]
+#[derive(Debug)]
+pub struct napi_node_version {
+  pub major: u32,
+  pub minor: u32,
+  pub patch: u32,
+  pub release: *const c_char,
+}
+
+#[repr(C)]
+#[derive(Debug)]
+/// Env that is shared between all contexts in same native module.
+pub struct EnvShared {
+  pub instance_data: *mut c_void,
+  pub data_finalize: Option<napi_finalize>,
+  pub data_finalize_hint: *mut c_void,
+  pub napi_wrap: v8::Global<v8::Private>,
+  pub finalize: Option<napi_finalize>,
+  pub finalize_hint: *mut c_void,
+  pub filename: *const c_char,
+}
+
+impl EnvShared {
+  pub fn new(napi_wrap: v8::Global<v8::Private>) -> Self {
+    Self {
+      instance_data: std::ptr::null_mut(),
+      data_finalize: None,
+      data_finalize_hint: std::ptr::null_mut(),
+      napi_wrap,
+      finalize: None,
+      finalize_hint: std::ptr::null_mut(),
+      filename: std::ptr::null(),
+    }
+  }
+}
+
+pub enum ThreadSafeFunctionStatus {
+  Alive,
+  Dead,
+}
+
+#[repr(C)]
+pub struct Env {
+  // pub struct Env<'a, 'b> {
+  // pub scope: &'a mut v8::HandleScope<'b>,
+  context: NonNull<v8::Context>,
+  pub isolate_ptr: *mut v8::OwnedIsolate,
+  pub open_handle_scopes: usize,
+  pub shared: *mut EnvShared,
+  pub async_work_sender: mpsc::UnboundedSender<PendingNapiAsyncWork>,
+  pub threadsafe_function_sender:
+    mpsc::UnboundedSender<ThreadSafeFunctionStatus>,
+}
+
+// unsafe impl Send for Env<'_, '_> {}
+// unsafe impl Sync for Env<'_, '_> {}
+unsafe impl Send for Env {}
+unsafe impl Sync for Env {}
+
+// impl<'a, 'b> Env<'a, 'b> {
+impl Env {
+  pub fn new(
+    isolate_ptr: *mut v8::OwnedIsolate,
+    // scope: &'a mut v8::HandleScope<'b>,
+    context: v8::Global<v8::Context>,
+    sender: mpsc::UnboundedSender<PendingNapiAsyncWork>,
+    threadsafe_function_sender: mpsc::UnboundedSender<ThreadSafeFunctionStatus>,
+  ) -> Self {
+    let sc = sender.clone();
+    ASYNC_WORK_SENDER.with(|s| {
+      s.replace(Some(sc));
+    });
+    let ts = threadsafe_function_sender.clone();
+    THREAD_SAFE_FN_SENDER.with(|s| {
+      s.replace(Some(ts));
+    });
+
+    Self {
+      // scope,
+      isolate_ptr,
+      context: context.into_raw(),
+      shared: std::ptr::null_mut(),
+      open_handle_scopes: 0,
+      async_work_sender: sender,
+      threadsafe_function_sender,
+    }
+  }
+
+  // TODO(@littledivy): Do we even need this?
```

Comment:
Good question, what is the use case?

Reply:
It was used to create a new Env from a new scope and an existing Env, but now that `Env`'s are independent of scopes this can be removed